File: python/Q460_lfu_cache.py
#
# @lc app=leetcode id=460 lang=python3
#
# [460] LFU Cache
#

# @lc code=start
class Node:
    def __init__(self, key: int, value: int) -> None:
        self.key: int = key
        self.value: int = value
        self.count: int = 1

        self.previous: Node | None = None
        self.next: Node | None = None


class LinkedList:
    def __init__(self) -> None:
        self.start = Node(-1, -1)
        self.end = Node(-1, -1)

        self.start.next = self.end
        self.end.previous = self.start

        self.size = 0

    def add_to_head(self, node: Node) -> None:
        tmp = self.start.next
        assert tmp is not None

        self.start.next = node
        node.next = tmp

        node.previous = self.start
        tmp.previous = node

        self.size += 1

    def remove_node(self, node: Node) -> None:
        previous = node.previous
        next = node.next

        assert previous is not None
        assert next is not None

        previous.next = next
        next.previous = previous

        self.size -= 1

    def remove_from_tail(self) -> Node:
        node = self.end.previous
        assert node is not None

        self.remove_node(node)

        return node


class LFUCache:
    def __init__(self, capacity: int) -> None:
        self.capacity = capacity

        # key -> Node
        self.dict: dict[int, Node] = dict()

        # f -> LinkedList
        self.mapping: dict[int, LinkedList] = dict()

        self.min_f: int = 0

    def update_f(self, node: Node) -> None:
        self.mapping[node.count].remove_node(node)

        node.count += 1

        if node.count not in self.mapping:
            self.mapping[node.count] = LinkedList()
        self.mapping[node.count].add_to_head(node)

        if node.count == self.min_f + 1 and self.mapping[self.min_f].size == 0:
            self.min_f += 1

    def get(self, key: int) -> int:
        if key not in self.dict:
            return -1

        node = self.dict[key]
        self.update_f(node)

        return node.value

    def put(self, key: int, value: int) -> None:
        # key already exist
        if key in self.dict:
            node = self.dict[key]
            node.value = value
            self.update_f(node)
            return

        # remove LRU node if capacity is full
        if len(self.dict) == self.capacity:
            removed_node = self.mapping[self.min_f].remove_from_tail()
            del self.dict[removed_node.key]

        new_node = Node(key, value)
        self.dict[key] = new_node
        if 1 not in self.mapping:
            self.mapping[1] = LinkedList()

        self.mapping[1].add_to_head(new_node)
        self.min_f = 1


# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end

File: rust/src/leetcode/Q460_lfu_cache.rs
/*
 * @lc app=leetcode id=460 lang=rust
 *
 * [460] LFU Cache
 */

// @lc code=start
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

struct Node {
    key: i32,
    value: i32,
    count: i32,
    previous: Option<Rc<RefCell<Node>>>,
    next: Option<Rc<RefCell<Node>>>,
}

impl Node {
    fn new(key: i32, value: i32) -> Self {
        Node {
            key,
            value,
            count: 1,
            previous: None,
            next: None,
        }
    }
}

struct LinkedList {
    start: Rc<RefCell<Node>>,
    end: Rc<RefCell<Node>>,
    size: i32,
}

impl LinkedList {
    fn new() -> Self {
        let start: Rc<RefCell<Node>> = Rc::new(RefCell::new(Node::new(-1, -1)));
        let end: Rc<RefCell<Node>> = Rc::new(RefCell::new(Node::new(-1, -1)));

        start.borrow_mut().next = Some(end.clone());
        end.borrow_mut().previous = Some(start.clone());

        LinkedList {
            start,
            end,
            size: 0,
        }
    }

    fn add_to_head(&mut self, node: Rc<RefCell<Node>>) {
        let tmp = self.start.borrow().next.clone().unwrap();

        self.start.borrow_mut().next = Some(node.clone());
        node.borrow_mut().next = Some(tmp.clone());

        node.borrow_mut().previous = Some(self.start.clone());
        tmp.borrow_mut().previous = Some(node.clone());

        self.size += 1;
    }

    fn remove_node(&mut self, node: Rc<RefCell<Node>>) {
        let previous = node.borrow().previous.clone().unwrap();
        let next = node.borrow().next.clone().unwrap();

        previous.borrow_mut().next = Some(next.clone());
        next.borrow_mut().previous = Some(previous.clone());

        self.size -= 1;
    }

    fn remove_from_tail(&mut self) -> Rc<RefCell<Node>> {
        let node = self.end.borrow().previous.clone().unwrap();

        self.remove_node(node.clone());

        node
    }
}

struct LFUCache {
    capacity: i32,
    node_dict: HashMap<i32, Rc<RefCell<Node>>>,
    f_dict: HashMap<i32, LinkedList>,
    min_f: i32,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl LFUCache {
    fn new(capacity: i32) -> Self {
        LFUCache {
            capacity,
            node_dict: HashMap::new(),
            f_dict: HashMap::new(),
            min_f: 0,
        }
    }

    fn update(&mut self, node: Rc<RefCell<Node>>) {
        let current_count = node.borrow().count;

        self.f_dict
            .get_mut(&current_count)
            .unwrap()
            .remove_node(node.clone());

        node.borrow_mut().count += 1;

        self.f_dict
            .entry(current_count + 1)
            .or_insert(LinkedList::new())
            .add_to_head(node.clone());

        if current_count == self.min_f && self.f_dict.get(&self.min_f).unwrap().size == 0 {
            self.min_f += 1;
        }
    }

    fn get(&mut self, key: i32) -> i32 {
        if !self.node_dict.contains_key(&key) {
            return -1;
        }

        let node = self.node_dict.get(&key).cloned().unwrap();
        self.update(node.clone());

        return node.borrow().value;
    }

    fn put(&mut self, key: i32, value: i32) {
        if self.capacity == 0 {
            return;
        }

        if let Some(node) = self.node_dict.get(&key).cloned() {
            node.borrow_mut().value = value;
            self.update(node);
            return;
        }

        if self.node_dict.len() as i32 == self.capacity {
            let list = self.f_dict.get_mut(&self.min_f).unwrap();
            let removed_node = list.remove_from_tail();
            self.node_dict.remove(&removed_node.borrow().key);
        }

        let new_node = Rc::new(RefCell::new(Node::new(key, value)));
        self.node_dict.insert(key, new_node.clone());
        self.f_dict
            .entry(1)
            .or_insert(LinkedList::new())
            .add_to_head(new_node);

        self.min_f = 1;
    }
}

/*
 * Your LFUCache object will be instantiated and called as such:
 * let obj = LFUCache::new(capacity);
 * let ret_1: i32 = obj.get(key);
 * obj.put(key, value);
 */
// @lc code=end
